duplicate.py

```diff
@@ -0,0 +1,140 @@
+# -*- coding: utf-8 -*-
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this file,
+# You can obtain one at http://mozilla.org/MPL/2.0/.
+
+# Using latent semantic indexing and similarity matrix
+
+import logging
+from collections import defaultdict
+
+import nltk
+from gensim import models, similarities
+from gensim.corpora import Dictionary
+from nltk.corpus import stopwords
+from nltk.stem.porter import PorterStemmer
+
+from bugbug import bugzilla
+
+logging.basicConfig(
+    format="%(asctime)s : %(levelname)s : %(message)s", level=logging.INFO
+)
+
+nltk.download("stopwords")
+
+# A map from bug id to its duplicate ids
+all_ids = defaultdict(int)
+duplicates = defaultdict(set)
+
+for bug in bugzilla.get_bugs():
+    all_ids[bug["id"]] = 1
+
+for bug in bugzilla.get_bugs():
+    dupes = set([bugs for bugs in bug["duplicates"] if all_ids[bugs] != 0])
+    if all_ids[bug["dupe_of"]] != 0:
+        dupes.add(bug["dupe_of"])
+
+    duplicates[bug["id"]] |= dupes
+    for dupe in dupes:
+        duplicates[dupe].add(bug["id"])
+
+
+class SimilarityLsi:
+    def __init__(self):
+
+        self.corpus = []
+
+        for bug in bugzilla.get_bugs():
+            self.corpus.append([bug["id"], bug["summary"]])
+
+        # cleaning the text
+        self.ps = PorterStemmer()
+        texts = [
+            [
+                self.ps.stem(word)
+                for word in summary.lower().split()
+                if word not in set(stopwords.words("english"))
+            ]
+            for bug_id, summary in self.corpus
+        ]
+
+        # Assigning unique integer ids to all words
+        self.dictionary = Dictionary(texts)
+
+        # conversion to bow
+        corpus_final = [self.dictionary.doc2bow(text) for text in texts]
+
+        # initializing and applying the tfidf transformation model on same corpus,resultant corpus is of same dimensions
+        tfidf = models.TfidfModel(corpus_final)
+        corpus_tfidf = tfidf[corpus_final]
+
+        # transform tfidf corpus to latent 300-D space via LATENT SEMANTIC INDEXING
+        self.lsi = models.LsiModel(
+            corpus_tfidf, id2word=self.dictionary, num_topics=300
+        )
+        corpus_lsi = self.lsi[corpus_tfidf]
+
+        # indexing the corpus
+        self.index = similarities.Similarity(
+            output_prefix="simdata.shdat", corpus=corpus_lsi, num_features=300
```

Comment:
@marco-c when this runs it creates 4 files named simdatax.shdat in the same destination, do we have problem with that?

Reply:
Yeah, that's fine!